pkg/api/subscribeWorker.go

```diff
@@ -24,7 +23,56 @@ const (
 	maxTopicLength            = 128
 )
 
-type listener = chan<- []*message_api.OriginatorEnvelope
+type listener struct {
+	ctx         context.Context
+	ch          chan<- []*message_api.OriginatorEnvelope
+	closed      bool
+	topics      map[string]struct{}
+	originators map[uint32]struct{}
+	isGlobal    bool
+}
+
+type listenerSet struct {
+	sync.Map // map[*listener]struct{}
+}
+
+func (ls *listenerSet) addListener(l *listener) {
+	ls.Store(l, struct{}{})
+}
+
+func (ls *listenerSet) removeListener(l *listener) {
+	ls.Delete(l)
+}
+
+func (ls *listenerSet) IsEmpty() bool {
+	empty := true
+	ls.Range(func(_, _ interface{}) bool {
+		empty = false
+		return false // stop iteration
+	})
+	return empty
+}
+
+// Maps from a key to a set of listeners
+type listenersMap[K comparable] struct {
+	sync.Map // map[K]*listenerSet
+}
+
+func (lm *listenersMap[K]) addListener(key K, l *listener) {
+	value, _ := lm.LoadOrStore(key, &listenerSet{})
+	set := value.(*listenerSet)
+	set.addListener(l)
+}
+
+func (lm *listenersMap[K]) removeListener(key K, l *listener) {
+	value, ok := lm.Load(key)
+	if !ok || value == nil {
+		return // Key doesn't exist, nothing to do
+	}
+	set := value.(*listenerSet)
+	set.removeListener(l)
+	// TODO(rich): Delete keys that hold empty sets
```

Comment:
The only thing I haven't been able to solve with this approach is here. I haven't worked out a way to delete keys for which there are no more listeners, in a way that is thread-safe and doesn't compromise readability or performance.

The issue is that checking that the set is empty, and then deleting it, is not an atomic operation. And I suspect that using something like `sync.map.CompareAndDelete()` won't work, because that is likely doing pointer comparison.

Any suggestions are welcome, I can also keep thinking about this - one approach would be a granular map of RWMutex, another approach would be to stop storing these listener lookups and iterate through all listeners on every payload. I'm assuming this has to be solved, otherwise we end up holding all topics in memory on the server which I assume is too much.

Reply:
I'm leaning towards using a channel here. Moving PR back to draft for now while I investigate that approach.